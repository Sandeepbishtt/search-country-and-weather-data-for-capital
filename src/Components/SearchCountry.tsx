import React, { useState, ChangeEvent } from "react";
import { Box, Input, Button } from "@mui/material";
import NoteAddIcon from "@material-ui/icons/NoteAdd";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router";
import axios from "axios";
import {
  addData,
  hideLoading,
  loading,
  showDetailReducer,
  showLoading,
} from "../Redux/SliceReducer";
import { CircularProgress } from "@mui/material";

const classes = {
  form: {
    marginLeft: "1.1rem",
    marginTop: "2rem",
  },
  Input: {
    width: "25rem",
    marginBottom: "1rem",
  },
  button :{
    marginTop:'1rem',
    marginLeft:'9rem'
  },
  loadStyle :{
    marginLeft:'50%',
    marginTop :'15rem'
  }

};

export const InputForm: React.FC = () => {
  const history = useHistory();
  const [input, setInput] = useState("");
  const dispatch = useDispatch();
  const load: boolean = useSelector(loading);

  const submitHandler = (e: React.FormEvent) => {
    e.preventDefault();
    getCountryData(input);
    setInput("");
  };

  const getCountryData = async (id: any) => {
    dispatch(showLoading());
    try {
      const response: any = await axios.get(
        `https://restcountries.com/v2/name/${input.toLowerCase().trim()}`
      );
      if (response.status === 200) {
        dispatch(addData(response.data));
        dispatch(showDetailReducer());
        dispatch(hideLoading());
        if (response.data.status !== 404) {
          history.push({
            pathname: "/CountryDetail",
          });
        } else {
          alert("Oops no country found!!!");
        }
      }
    } catch (err) {
      throw new Error("Cannot Fetch Country");
    }
  };

  return (
    <>
      {load ? (
        <CircularProgress
          size={68}
          style={classes.loadStyle}
        />
      ) : (
        <Box
          border={1}
          borderColor="grey.500"
          borderRadius={5}
          bgcolor="white.main"
          boxShadow={3}
          width={450}
          height={150}
          mx="auto"
          mt="2rem"
        >
          <form data-test='login-form' style={classes.form} onSubmit={submitHandler}>
            <Input
              style={classes.Input}
              type="text"
              required
              name='countryName'
              className="form-control"
              placeholder="Enter Country"
              onChange={(event: ChangeEvent<HTMLInputElement>) =>
                setInput(event.target.value)
              }
              value={input}
            />
            <Button
              style={classes.button}
              type="submit"
              variant="contained"
              color="primary"
              startIcon={<NoteAddIcon />}
              disabled={input === ""}
            >
              Submit
            </Button>
          </form>
        </Box>
      )}
    </>
  );
};

export default InputForm;
