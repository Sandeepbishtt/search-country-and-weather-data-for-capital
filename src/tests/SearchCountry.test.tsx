import * as ReactDOM from "react-dom";
import { Provider } from "react-redux";
import store from "../Redux/Store";
import React from "react";
import SearchCountry from '../Components/SearchCountry'
import "@testing-library/jest-dom";
import { BrowserRouter } from "react-router-dom";
import {fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';


describe("Home Component tests", () => {
  let container: HTMLDivElement;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
    ReactDOM.render(
      <React.StrictMode>
        <BrowserRouter>
          <Provider store={store}>
            <SearchCountry />
          </Provider>
        </BrowserRouter>
      </React.StrictMode>,
      container
    );
  });
  afterEach(() => {
    document.body.removeChild(container);
    container.remove();
  });
  it("Render correctly initial document", () => {
    expect(container).toMatchSnapshot();
  });
  it('Redirect to correct location', () => {
    const button = container.querySelectorAll('button');
    const loginButton = button[0];
    fireEvent.click(loginButton);
    expect(window.location.pathname).toEqual('/')
});

it('checking input in Component', () => {
  const inputs = container.querySelectorAll('input');
  expect(inputs).toHaveLength(1);
})

it('checking button in component', () => {
  const buttons = container.querySelectorAll('button');
  expect(buttons).toHaveLength(1);
})


it('Renders correctly initial document with data-test query', () => {
  expect(container.querySelector("[data-test='login-form']")).toBeInTheDocument();
  expect(container.querySelector("input")?.getAttribute('name'))
      .toBe('countryName');
});

});
