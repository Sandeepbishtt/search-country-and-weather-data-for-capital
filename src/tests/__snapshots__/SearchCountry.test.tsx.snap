// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component tests Render correctly initial document 1`] = `
<div>
  <div
    class="MuiBox-root css-1niv0cm"
  >
    <form
      data-test="login-form"
      style="margin-left: 1.1rem; margin-top: 2rem;"
    >
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary form-control css-a88p61-MuiInputBase-root-MuiInput-root"
        required=""
        style="width: 25rem; margin-bottom: 1rem;"
      >
        <input
          class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          name="countryName"
          placeholder="Enter Country"
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled=""
        style="margin-top: 1rem; margin-left: 9rem;"
        tabindex="-1"
        type="submit"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 14h-3v3h-2v-3H8v-2h3v-3h2v3h3v2zm-3-7V3.5L18.5 9H13z"
            />
          </svg>
        </span>
        Submit
      </button>
    </form>
  </div>
</div>
`;
